Red 3.4.1 - Changelog (#4193)
Cog-Creators/Red-DiscordBot: docs/changelog_3_4_0.rst

Mod
***

- Fixed nicknames not being properly stored and logged (:issue:`4131`)
- Added ``[p]modset mentionspam strict`` allowing for duplicated mentions to count to mention spam cap (:issue:`4359`)
- Fixed plural typos in ``[p]userinfo`` (:issue:`4397`, :issue:`4379`)

Trivia Lists
************

- Fixed incorrect order of Machamp and Machoke questions (:issue:`4424`)
- Added new MLB trivia list (:issue:`4455`)

Warnings
********

- Fixed users being able to warn users above them in hierarchy (:issue:`4100`)
- Added bool arguments to toggle commands to improve consistency (:issue:`4409`)

Developer changelog
-------------------

Core
****

- Added ``bot.remove_shared_api_services`` to remove all keys and tokens associated with an API service (:issue:`4370`)
- Added option to return all tokens for an API service if ``service_name`` is not specified in ``bot.get_shared_api_tokens`` (:issue:`4370`)

Bank
****

- Bank API methods now throw TypeError if a non-integer amount is supplied (:issue:`4376`)

Mod
***

- Deprecated ``redbot.core.utils.mod.is_allowed_by_hierarchy`` (:issue:`4435`)

Modlog
******

- Added option to accept a ``discord.Object`` in case creation (:issue:`4326`)

> - Added ``last_known_username`` parameter to ``create_case`` function (:issue:`4326`)
> - Fixed error being raised with a deleted channel in ``case.message_content`` (:issue:`4415`)

Reviewer: These should link to relevant docs:
```suggestion
- Added ``last_known_username`` parameter to `modlog.create_case()` function (:issue:`4326`)
- Fixed error being raised with a deleted channel in `Case.message_content()` (:issue:`4415`)
```

Utility
*******

> - Added ``redbot.core.utils.get_end_user_statement`` to attempt to fetch a cog's End User Data Statement (:issue:`4404`)
> - Added ``redbot.core.utils.chat_formatting.quote`` to quote text in a message (:issue:`4425`)

Reviewer: These should link to relevant docs:
```suggestion
- Added `redbot.core.utils.get_end_user_statement()` to attempt to fetch a cog's End User Data Statement (:issue:`4404`)
- Added `redbot.core.utils.chat_formatting.quote()` to quote text in a message (:issue:`4425`)
```

Documentation changes
---------------------

Config
******

- Added custom group documentation and tutorial (:issue:`4416`, :issue:`2896`)

Modlog
******

> - Clarified that naive ``datetime`` objects will be treated as local times for paramaters ``created_at`` and ``until`` in ``create_case`` (:issue:`4389`)

Reviewer: This should link to relevant docs:
```suggestion
- Clarified that naive ``datetime`` objects will be treated as local times for parameters ``created_at`` and ``until`` in `modlog.create_case()` (:issue:`4389`)
```
Also, I found a typo so I included a fix for that in the suggestion above.

- Fixed error being raised with ``[p]set nickname`` when no nickname was provided (:issue:`4451`)
- Fixed and clarified errors being raised with ``[p]set username`` (:issue:`4463`)
- Fixed ungraceful error raising when ``[p]unload`` output is larger than 2k characters (:issue:`4469`)
- Fixed ungraceful error raising when providing ``[p]choose`` with empty options (:issue:`4499`)
- Added ``[p]set api list`` to list all currently set API services, without tokens (:issue:`4370`)
- Added ``[p]set api remove`` to remove API services, including tokens (:issue:`4370`)
- Added ``[p]helpset usetick``, toggling command message being ticked when help is sent to DM (:issue:`4467`, :issue:`4075`)
- Added default color field to ``[p]set showsettings`` (:issue:`4498`, :issue:`4497`)

Audio
*****

- Fixed scattered grammar and typo fixes (:issue:`4446`)
- Fixed Bandcamp playback (:issue:`4504`)
- Fixed YouTube playlist playback (:issue:`4504`)
- Fixed YouTube searching issues (:issue:`4504`)
- Fixed YouTube age restricted track playback (:issue:`4504`)
- Removed lavalink logs from being added to backup (:issue:`4453`, :issue:`4452`)

> - Added Global Audio API, to cut down on Youtube 429 errors (:issue:`4446`)

Reviewer 2: ```suggestion
- Added Global Audio API, to cut down on Youtube 429 errors and allow Spotify playback past users quota. (:issue:`4446`)
```

> - Added persistent queues, allowing for queues to ne restored after being closed (:issue:`4446`)

Reviewer 2: ```suggestion
- Added persistent queues, allowing for queues to ne restored after on a bot restart or cog reload (:issue:`4446`)
```

- Added ``[p]audioset restart``, allowing for Lavalink connection to be restarted (:issue:`4446`)
- Added ``[p]audioset autodeafen``, allowing for bot to auto-deafen itself when entering voice channel (:issue:`4446`)

> - Added Spotify search locale per user (:issue:`4446`)

Reviewer 2: ```suggestion
- Added ``[p]audioset mycountrycode``. allowing Spotify search locale per user (:issue:`4446`)
```

.. 3.4.x Changelogs

Redbot 3.4.1 (2020-10-27)
=========================

| Thanks to all these amazing people that contributed to this release:
| :ghuser:`absj30`, :ghuser:`aikaterna`, :ghuser:`bobloy`, :ghuser:`chloecormier`, :ghuser:`Dav-Git`, :ghuser:`Drapersniper`, :ghuser:`fixator10`, :ghuser:`Flame442`, :ghuser:`flaree`, :ghuser:`Generaleoley`, :ghuser:`hisztendahl`, :ghuser:`jack1142`, :ghuser:`KaiGucci`, :ghuser:`Kowlin`, :ghuser:`maxbooiii`, :ghuser:`MeatyChunks`, :ghuser:`NeuroAssassin`, :ghuser:`nfitzen`, :ghuser:`palmtree5`, :ghuser:`phenom4n4n`, :ghuser:`PredaaA`, :ghuser:`Predeactor`, :ghuser:`PythonTryHard`, :ghuser:`SharkyTheKing`, :ghuser:`Stonedestroyer`, :ghuser:`thisisjvgrace`, :ghuser:`TrustyJAID`, :ghuser:`TurnrDev`, :ghuser:`Vexed01`, :ghuser:`Vuks69`, :ghuser:`xBlynd`, :ghuser:`zephyrkul`
|
| **Read before updating**:
| 1. This Red update bumps discord.py to version 1.5.1, which explicitly requests Discord intents. Red requires all Prvileged Intents to be enabled. More information can be found at :ref:`enabling-privileged-intents`.

> | 2. Mutes functionality has been moved from Mod cog to a new separate cog (Mutes) featuring timed and role-based mutes. If you were using it (or want to start now), you can load the new cog with ``[p]load mutes``. You can see full `Mutes changelog below <important-341-1>`.

Reviewer 3: ```suggestion
| 2. Mutes functionality has been moved from the Mod cog to a new separate cog (Mutes) featuring timed and role-based mutes. If you were using it (or want to start now), you can load the new cog with ``[p]load mutes``. You can see the full `Mutes changelog below <important-341-1>`.
```

> | 3. Information for Audio users that are using external Lavalink instance (if you don't know what that is, you should skip this point):

Reviewer 3: ```suggestion
| 3. Information for Audio users that are using an external Lavalink instance (if you don't know what that is, you should skip this point):
```

| We've updated our ``application.yml`` file and you should update your instance's ``application.yml`` appropriately.
  Please ensure that the WS port in Audio's settings (``[p]llset wsport``) is set to the port from the ``application.yml``.

End-user changelog
------------------

Core Bot
********

> - Fixed incorrect error being reported on ``[p]set name`` when the passed name was longer than 32 characters (:issue:`4364`, :issue:`4363`)

Reviewer 3: ```suggestion
- Fixed an incorrect error being reported on ``[p]set name`` when the passed name was longer than 32 characters (:issue:`4364`, :issue:`4363`)
```

- Fixed ``[p]set nickname`` erroring when the passed name was longer than 32 characters (:issue:`4364`, :issue:`4363`)
- Fixed an ungraceful error being raised when running ``[p]traceback`` with closed DMs (:issue:`4329`)
- Fixed errors that could arise from invalid URLs in ``[p]set avatar`` (:issue:`4437`)

> - Fixed error being raised with ``[p]set nickname`` when no nickname was provided (:issue:`4451`)

Reviewer 3: ```suggestion
- Fixed an error being raised with ``[p]set nickname`` when no nickname was provided (:issue:`4451`)
```

- Fixed and clarified errors being raised with ``[p]set username`` (:issue:`4463`)

> - Fixed ungraceful error raising when ``[p]unload`` output is larger than 2k characters (:issue:`4469`)
> - Fixed ungraceful error raising when providing ``[p]choose`` with empty options (:issue:`4499`)
> - Fixed ungraceful error raising when a bot left a guild while a menu was open (:issue:`3902`)

Reviewer 3: ```suggestion
- Fixed an ungraceful error being raised when the output of ``[p]unload`` is larger than 2k characters (:issue:`4469`)
- Fixed an ungraceful error being raised when running ``[p]choose`` with empty options (:issue:`4499`)
- Fixed an ungraceful error being raised when a bot left a guild while a menu was open (:issue:`3902`)
```

> - Fixed info missing on non-embed version of ``[p]debuginfo`` (:issue:`4524`)

Reviewer 3: ```suggestion
- Fixed info missing on the non-embed version of ``[p]debuginfo`` (:issue:`4524`)
```

- Added ``[p]set api list`` to list all currently set API services, without tokens (:issue:`4370`)
- Added ``[p]set api remove`` to remove API services, including tokens (:issue:`4370`)
- Added ``[p]helpset usetick``, toggling command message being ticked when help is sent to DM (:issue:`4467`, :issue:`4075`)

> - Added default color field to ``[p]set showsettings`` (:issue:`4498`, :issue:`4497`)

Reviewer 3: ```suggestion
- Added a default color field to ``[p]set showsettings`` (:issue:`4498`, :issue:`4497`)
```

> - Added datapath and metadata file to ``[p]debuginfo`` (:issue:`4524`)

Reviewer 3: ```suggestion
- Added the datapath and metadata file to ``[p]debuginfo`` (:issue:`4524`)
```

> - Added list of disabled intents to ``[p]debuginfo`` (:issue:`4423`)

Reviewer 3: ```suggestion
- Added a list of disabled intents to ``[p]debuginfo`` (:issue:`4423`)
```

- Bumped discord.py dependency to version 1.5.1 (:issue:`4423`)
- Locales and regional formats can now be set in individual guilds using ``[p]set locale`` and ``[p]set regionalformat`` (:issue:`3896`, :issue:`1970`)

    - Global locale and regional format setters have been renamed to ``[p]set globallocale`` and ``[p]set globalregionalformat``

Audio
*****

> - Fixed scattered grammar and typo fixes (:issue:`4446`)

Reviewer 3: ```suggestion
- Scattered grammar and typo fixes (:issue:`4446`)
```

Reviewer: That changes the meaning. This fixed the grammar being scattered. It did not scatter the grammer.

- Fixed Bandcamp playback (:issue:`4504`)
- Fixed YouTube playlist playback (:issue:`4504`)
- Fixed YouTube searching issues (:issue:`4504`)
- Fixed YouTube age restricted track playback (:issue:`4504`)

> - Fixed Audio cog not being translated when setting locale (:issue:`4492`, :issue:`4495`)

Reviewer 3: ```suggestion
- Fixed the Audio cog not being translated when setting locale (:issue:`4492`, :issue:`4495`)
```

- Fixed tracks getting stuck at 0:00 after long player sessions (:issue:`4529`)
- Removed lavalink logs from being added to backup (:issue:`4453`, :issue:`4452`)
- Removed stream durations from being in queue duration (:issue:`4513`)

> - Added Global Audio API, to cut down on Youtube 429 errors and allow Spotify playback past users quota. (:issue:`4446`)

Reviewer 3: ```suggestion
- Added the Global Audio API, to cut down on Youtube 429 errors and allow Spotify playback past user's quota. (:issue:`4446`)
```

- Added persistent queues, allowing for queues to be restored on a bot restart or cog reload (:issue:`4446`)
- Added ``[p]audioset restart``, allowing for Lavalink connection to be restarted (:issue:`4446`)
- Added ``[p]audioset autodeafen``, allowing for bot to auto-deafen itself when entering voice channel (:issue:`4446`)
- Added ``[p]audioset mycountrycode``, allowing Spotify search locale per user (:issue:`4446`)

> - Added ``[p]llsetup java``, allowing for custom Java executable path (:issue:`4446`)

Reviewer 3: ```suggestion
- Added ``[p]llsetup java``, allowing for a custom Java executable path (:issue:`4446`)
```

- Added ``[p]llset info`` to show Lavalink settings (:issue:`4527`)

> - Added ``[p]audioset logs`` to download Lavalink logs if Lavalink server is set to internal (:issue:`4527`)

Reviewer 3: ```suggestion
- Added ``[p]audioset logs`` to download Lavalink logs if the Lavalink server is set to internal (:issue:`4527`)
```

Cleanup
*******

- Allowed ``[p]cleanup self`` to work in DMs for all users (:issue:`4481`)

Custom Commands
***************

> - Fixed ungraceful error being thrown on ``[p]cc edit`` (:issue:`4325`)

Reviewer 3: ```suggestion
- Fixed an ungraceful error being thrown on ``[p]cc edit`` (:issue:`4325`)
```

Dev
***

- Added ``[p]repl pause`` to pause/resume the REPL session in the current channel (:issue:`4366`)

Economy
*******

> - Added embed option for ``[p]leaderboard`` (:issue:`4184`, :issue:`4104`)

Reviewer 3: ```suggestion
- Added an embed option for ``[p]leaderboard`` (:issue:`4184`, :issue:`4104`)
```

General
*******

- Fixed issues with text not being properly URL encoded (:issue:`4024`)

> - Fixed ungraceful error occuring when title was longer than 256 characters in ``[p]urban`` (:issue:`4474`)

Reviewer 3: ```suggestion
- Fixed an ungraceful error occurring when a title is longer than 256 characters in ``[p]urban`` (:issue:`4474`)
```

- Changed "boosters" to "boosts" in ``[p]serverinfo`` to clarify what the number represents (:issue:`4507`)

Mod
***

> - Added ``[p]modset mentionspam strict`` allowing for duplicated mentions to count to mention spam cap (:issue:`4359`)

Reviewer 3: ```suggestion
- Added ``[p]modset mentionspam strict`` allowing for duplicated mentions to count towards the mention spam cap (:issue:`4359`)
```

> - Added option to ban users not in the guild to ``[p]ban`` (:issue:`4422`, :issue:`4419`)

Reviewer 3: ```suggestion
- Added an option to ban users not in the guild to ``[p]ban`` (:issue:`4422`, :issue:`4419`)
```

> - Added default tempban duration for ``[p]tempban`` (:issue:`4473`, :issue:`3992`)

Reviewer 3: ```suggestion
- Added a default tempban duration for ``[p]tempban`` (:issue:`4473`, :issue:`3992`)
```

- Fixed nicknames not being properly stored and logged (:issue:`4131`)
- Fixed plural typos in ``[p]userinfo`` (:issue:`4397`, :issue:`4379`)
- Renamed ``[p]hackban`` to ``[p]massban``, keeping ``[p]hackban`` as an alias, allowing for multiple users to be banned at once (:issue:`4422`, :issue:`4419`)
- Moved mutes to a separate, individual cog (:issue:`3634`)

.. _important-341-1:

Mutes
*****

- Added ``[p]muteset forcerole`` to make mutes role based, instead of permission based (:issue:`3634`)
- Added ``[p]mutechannel`` to mute a user in the current channel (:issue:`3634`)
- Added an optional time argument to all mutes, to specify when the user should be unmuted (:issue:`3634`)
- Mute commands can now take multiple user arguments, to mute multiple users at a time (:issue:`3634`)
- Changed ``[p]mute`` to default to serverwide muting (:issue:`3634`)

Modlog
******

> - Fixed error being raised when running ``[p]casesfor`` and ``[p]case`` (:issue:`4415`)

Reviewer 3: ```suggestion
- Fixed an error being raised when running ``[p]casesfor`` and ``[p]case`` (:issue:`4415`)
```

- Long reasons in Modlog are now properly shortened in message content (:issue:`4541`)

Trivia Lists
************

- Fixed incorrect order of Machamp and Machoke questions (:issue:`4424`)
- Added new MLB trivia list (:issue:`4455`)
- Added new Who's That Pokémon - Gen. IV trivia list (:issue:`4434`)
- Added new Hockey trivia list (:issue:`4384`)

Warnings
********

- Fixed users being able to warn users above them in hierarchy (:issue:`4100`)
- Added bool arguments to toggle commands to improve consistency (:issue:`4409`)

Developer changelog
-------------------

| **Important:**

> | 1. Red now allows to set locale per guild, which requires 3rd-party cogs to set contextual locale manually in code ran outside of command's context. See the `Core Bot changelog below <important-dev-341-1>` for more information.

Reviewer 3: ```suggestion
| 1. Red now allows users to set locale per guild, which requires 3rd-party cogs to set contextual locale manually in code ran outside of command's context. See the `Core Bot changelog below <important-dev-341-1>` for more information.
```

.. _important-dev-341-1:

Core Bot
********

- Added API for setting contextual locales (:issue:`3896`, :issue:`1970`)

    - New function added: `redbot.core.i18n.set_contextual_locales_from_guild()`
    - Contextual locale is automatically set for commands and only needs to be done manually for things like event listeners; see `guidelines-for-cog-creators` for more information

- Added `bot.remove_shared_api_services() <RedBase.remove_shared_api_services()>` to remove all keys and tokens associated with an API service (:issue:`4370`)

> - Added option to return all tokens for an API service if ``service_name`` is not specified in `bot.get_shared_api_tokens() <RedBase.get_shared_api_tokens()>` (:issue:`4370`)

Reviewer 3: ```suggestion
- Added an option to return all tokens for an API service if ``service_name`` is not specified in `bot.get_shared_api_tokens() <RedBase.get_shared_api_tokens()>` (:issue:`4370`)
```

- Added `bot.get_or_fetch_user() <RedBase.get_or_fetch_user()>` and `bot.get_or_fetch_member() <RedBase.get_or_fetch_member()>` methods (:issue:`4403`, :issue:`4402`)
- Moved ``redbot.core.checks.bot_in_a_guild()`` to `redbot.core.commands.bot_in_a_guild()` (old name has been left as an alias) (:issue:`4515`, :issue:`4510`)

Bank
****

- Bank API methods now consistently throw TypeError if a non-integer amount is supplied (:issue:`4376`)

Mod
***

- Deprecated ``redbot.core.utils.mod.is_allowed_by_hierarchy`` (:issue:`4435`)

Modlog
******

> - Added option to accept a ``discord.Object`` in case creation (:issue:`4326`)

Reviewer 3: ```suggestion
- Added an option to accept a ``discord.Object`` in case creation (:issue:`4326`)
```

- Added ``last_known_username`` parameter to `modlog.create_case()` function (:issue:`4326`)

> - Fixed error being raised with a deleted channel in `Case.message_content()` (:issue:`4415`)

Reviewer 3: ```suggestion
- Fixed an error being raised with a deleted channel in `Case.message_content()` (:issue:`4415`)
```

Utility
*******

- Added `redbot.core.utils.get_end_user_data_statement()` and `redbot.core.utils.get_end_user_data_statement_or_raise()` to attempt to fetch a cog's End User Data Statement (:issue:`4404`)
- Added `redbot.core.utils.chat_formatting.quote()` to quote text in a message (:issue:`4425`)

Documentation changes
---------------------

Config
******

- Added custom group documentation and tutorial (:issue:`4416`, :issue:`2896`)

Modlog
******

- Clarified that naive ``datetime`` objects will be treated as local times for parameters ``created_at`` and ``until`` in `modlog.create_case()` (:issue:`4389`)

Other
*****

- Added guide to creating a Bot Application in Discord Developer Portal, with enabling intents (:issue:`4502`)

Miscellaneous
-------------

- Added JSON schema files for ``info.json`` files (:issue:`4375`)
- Added ``[all]`` and ``[dev]`` bundled install extras (:issue:`4443`)

> - Replaced link to approved repository list on CogBoard and references to ``cogs.red`` with a link to new Red Index (:issue:`4439`)

Reviewer 3: ```suggestion
- Replaced the link to the approved repository list on CogBoard and references to ``cogs.red`` with a link to new Red Index (:issue:`4439`)
```

- Improved documentation about arguments in command syntax (:issue:`4058`)

> - Replaced a few instances of Red with bot name in command docstrings (:issue:`4470`)

Reviewer 3: ```suggestion
- Replaced a few instances of Red with the bot name in command docstrings (:issue:`4470`)
```

Redbot 3.4.1 (Unreleased)
=========================
| Thanks to all these amazing people that contributed to this release:

> | :ghuser:`fixator10`, :ghuser:`Flame442`, :ghuser:`Generaleoley`, :ghuser:`hisztendahl`, :ghuser:`jack1142`, :ghuser:`Kowlin`, :ghuser:`maxbooiii`, :ghuser:`NeuroAssassin`, :ghuser:`nfitzen`, :ghuser:`PythonTryHard`, :ghuser:`SharkyTheKing`, :ghuser:`StoneDestroyer`, :ghuser:`TrustyJAID`, :ghuser:`TurnrDev`, :ghuser:`Vuks69`, :ghuser:`xBlynd`, :ghuser:`zephyrkul`

Reviewer 4: Capital F
```suggestion
| :ghuser:`Fixator10`, :ghuser:`Flame442`, :ghuser:`Generaleoley`, :ghuser:`hisztendahl`, :ghuser:`jack1142`, :ghuser:`Kowlin`, :ghuser:`maxbooiii`, :ghuser:`NeuroAssassin`, :ghuser:`nfitzen`, :ghuser:`PythonTryHard`, :ghuser:`SharkyTheKing`, :ghuser:`StoneDestroyer`, :ghuser:`TrustyJAID`, :ghuser:`TurnrDev`, :ghuser:`Vuks69`, :ghuser:`xBlynd`, :ghuser:`zephyrkul`
```

- Fixed incorrect error being reported on ``[p]set name`` when the passed name was longer than 32 characters (:issue:`4364`, :issue:`4363`)
- Fixed ``[p]set nickname`` erroring when the passed name was longer than 32 characters (:issue:`4364`, :issue:`4363`)
- Fixed an ungraceful error being raised when running ``[p]traceback`` with closed DMs (:issue:`4329`)

> - Fixed errors that could arise from invalid URLs (:issue:`4437`)

Reviewer: ```suggestion
- Fixed errors that could arise from invalid URLs in ``[p]set avatar`` (:issue:`4437`)
```

Developer changelog
-------------------

Core
****

> - Added ``bot.remove_shared_api_services`` to remove all keys and tokens associated with an API service (:issue:`4370`)
> - Added option to return all tokens for an API service if ``service_name`` is not specified in ``bot.get_shared_api_tokens`` (:issue:`4370`)

Reviewer: These should link to relevant docs:
```suggestion
- Added `bot.remove_shared_api_services() <RedBase.remove_shared_api_services()>` to remove all keys and tokens associated with an API service (:issue:`4370`)
- Added option to return all tokens for an API service if ``service_name`` is not specified in `bot.get_shared_api_tokens() <RedBase.get_shared_api_tokens()>` (:issue:`4370`)
```

- Added option to accept a ``discord.Object`` in case creation (:issue:`4326`)
- Added ``last_known_username`` parameter to ``create_case`` function (:issue:`4326`)

Utility
*******

> - Added ``redbot.core.utils.get_end_user_statement`` to attempt to fetch a cog's End User Data Statement (:issue:`4404`)

Reviewer: a) the function is called `get_end_user_data_statement`
b) there was also `get_end_user_data_statement_or_raise` function added

- Added option to accept a ``discord.Object`` in case creation (:issue:`4326`)
- Added ``last_known_username`` parameter to ``create_case`` function (:issue:`4326`)

> - Fixed error being raised with a deleted channel in ``case.message_content`` (:issue:`4415`)

Reviewer: This also means we fixed end-user facing commands: `[p]casesfor` and `[p]case`

-------------

- Added JSON schema files for ``info.json`` files (:issue:`4375`)

> - Replaced ``cogs.red`` with new Red Index (:issue:`4439`)

Reviewer: We actually replaced the link to the approved repo list on CogBoard with Index, `cogs.red`-like service is still planned.
